examples/verified-fetch/fetcher.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return fmt.Errorf("failed to fetch CAR: %w", err)
+	}
+
+	rc, err = carToFileStream(ctx, rc, imPath)
+	if err != nil {
+		return fmt.Errorf("failed to read car stream: %w", err)
+	}
+
+	fd, err := os.OpenFile(output, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
+	if err != nil {
+		return fmt.Errorf("failed to open output file: %w", err)
+	}
+
+	_, err = io.Copy(fd, rc)
+	return err
+}
+
+func (f *fetcher) Close() error {
+	return nil
+}
+
+func (f *fetcher) resolvePath(ctx context.Context, p path.Path) (path.ImmutablePath, error) {
+	for p.Mutable() {
+		// Download IPNS record and verify through the gateway, or resolve the
+		// DNSLink with the default DNS resolver.
+		name, err := ipns.NameFromString(p.Segments()[1])
+		if err == nil {
+			p, err = f.resolveIPNS(ctx, name)
+		} else {
+			p, err = f.resolveDNSLink(ctx, p)
+		}
+
+		if err != nil {
+			return path.ImmutablePath{}, err
+		}
+	}
+
+	return path.NewImmutablePath(p)
+}
+
+func (f *fetcher) resolveIPNS(ctx context.Context, name ipns.Name) (path.Path, error) {
+	rc, err := f.httpRequest(ctx, name.AsPath(), "application/vnd.ipfs.ipns-record")
+	if err != nil {
+		return path.ImmutablePath{}, err
+	}
+
+	rc = newLimitReadCloser(rc, int64(ipns.MaxRecordSize))
+	rawRecord, err := io.ReadAll(rc)
+	if err != nil {
+		return path.ImmutablePath{}, err
+	}
+
+	rec, err := ipns.UnmarshalRecord(rawRecord)
+	if err != nil {
+		return path.ImmutablePath{}, err
+	}
+
+	err = ipns.ValidateWithName(rec, name)
+	if err != nil {
+		return path.ImmutablePath{}, err
+	}
+
+	return rec.Value()
+}
+
+func (f *fetcher) resolveDNSLink(ctx context.Context, p path.Path) (path.Path, error) {
+	dnsResolver := namesys.NewDNSResolver(madns.DefaultResolver.LookupTXT)
+	res, err := dnsResolver.Resolve(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	return res.Path, nil
+}
+
+func (f *fetcher) httpRequest(ctx context.Context, p path.Path, accept string) (io.ReadCloser, error) {
+	url := f.gateway + p.String()
+	fmt.Printf("Fetching with HTTP: %q\n", url)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("http.NewRequest error: %w", err)
+	}
+	req.Header.Set("Accept", accept)
+
+	if f.userAgent != "" {
+		req.Header.Set("User-Agent", f.userAgent)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("http.DefaultClient.Do error: %w", err)
+	}
+
+	if resp.StatusCode >= 400 {
+		defer resp.Body.Close()
+		mes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("error reading error body: %w", err)
+		}
+		return nil, fmt.Errorf("GET %s error: %s: %s", url, resp.Status, string(mes))
+	}
+
+	var rc io.ReadCloser
+	if f.limit > 0 {
+		rc = newLimitReadCloser(resp.Body, f.limit)
+	} else {
+		rc = resp.Body
+	}
+
+	return rc, nil
+}
+
+func carToFileStream(ctx context.Context, r io.ReadCloser, imPath path.ImmutablePath) (io.ReadCloser, error) {
```

Comment:
This doesn't do incremental verification from the gateways which might be fine, it depends what we're trying to do with this example. So not necessarily a blocker here but we should clarify what the plan is.

If we were trying to make this a library we should be able to turn on (or have by default) incremental verification. Code that does similar work should exist in #587, but at a high level:
1. If using DFS + duplicates: Have a blockstore that pulls data from the stream on demand and validates the multihash is the one being asked for, else it fails
2. If using DFS + no duplicates: Have a blockstore that pulls validated data from the stream (or the existing blockstore), if the requested block is neither the next block in the stream nor local then fail
3. If type is unknown either fail, or guess as 2

Reply:
If we were to merge #587 before, it would actually be easy to reuse the graph gateway backend: you could just instantiate it and call `GetCAR`, or even `Get` and stream directly to the file system. Leeching from the gateway code, but would do exactly what we want.

Reply:
@hacdias https://github.com/ipfs/boxo/pull/587 landed, would you be able to refactor this example to use that, and have incremental verification? 

(if so, fetch my changes before you start)

Reply:
@lidel in retrospective, I'm not sure if we have the best code to do this at the moment. Using `NewRemoteCarBackend` would make it very simple to resolve the given IPFS Path and downloading a CAR file. The problem then becomes extracting the CAR file. We don't have any good library to do it unfortunately. For that, I would need to resort to basically copying all of the code in this file: https://github.com/ipld/go-car/blob/master/cmd/car/extract.go

Or... we can make it such that this example simply downloads a CAR file and we can suggest the `go-car` CLI tool to then unpack the file. That would make it much simpler. What do you think?

Reply:
Yes! Will do :)